File: 1717_maximum_score_from_removing_substrings/Solution.java
// LeetCode 1717 - Maximum Score From Removing Substrings
// https://leetcode.com/problems/maximum-score-from-removing-substrings/

class Solution {
    public int maximumGain(String s, int x, int y) {
        int first;
        int second;
        StringBuilder rest = new StringBuilder();
        if (x >= y) {
            first = remove(s, 'a', 'b', x, rest);
            second = remove(rest.toString(), 'b', 'a', y, new StringBuilder());
        } else {
            first = remove(s, 'b', 'a', y, rest);
            second = remove(rest.toString(), 'a', 'b', x, new StringBuilder());
        }
        return first + second;
    }

    private int remove(String text, char open, char close, int score, StringBuilder rest) {
        int gained = 0;
        for (int i = 0; i < text.length(); i++) {
            char ch = text.charAt(i);
            if (rest.length() > 0 && rest.charAt(rest.length() - 1) == open && ch == close) {
                rest.deleteCharAt(rest.length() - 1);
                gained += score;
            } else {
                rest.append(ch);
            }
        }
        return gained;
    }
}

File: 1717_maximum_score_from_removing_substrings/Solution.kt
// LeetCode 1717 - Maximum Score From Removing Substrings
// https://leetcode.com/problems/maximum-score-from-removing-substrings/

class Solution {
    fun maximumGain(s: String, x: Int, y: Int): Int {
        fun remove(text: String, open: Char, close: Char, score: Int): Pair<String, Int> {
            val stack = StringBuilder()
            var gained = 0
            for (ch in text) {
                if (stack.isNotEmpty() && stack.last() == open && ch == close) {
                    stack.deleteCharAt(stack.length - 1)
                    gained += score
                } else {
                    stack.append(ch)
                }
            }
            return Pair(stack.toString(), gained)
        }

        return if (x >= y) {
            val (rest, first) = remove(s, 'a', 'b', x)
            val (_, second) = remove(rest, 'b', 'a', y)
            first + second
        } else {
            val (rest, first) = remove(s, 'b', 'a', y)
            val (_, second) = remove(rest, 'a', 'b', x)
            first + second
        }
    }
}
